File: frontend/src/components/app/workspace/action-registry.ts
import type { AppName } from "@/components/app-icon";

export type ActionType =
  | "send_email"
  | "send_bulk_email"
  | "create_calendar_event"
  | "create_notion_page";

export type FieldConfig =
  | {
      name: string;
      label: string;
      kind: "text" | "email" | "datetime-local";
      placeholder?: string;
    }
  | { name: string; label: string; kind: "textarea"; placeholder?: string }
  // Uploads and parses a CSV/XLSX recipient list — handled specially by
  // ActionForm, not a plain text input.
  | { name: string; label: string; kind: "recipients"; placeholder?: string };

export type ActionConfig = {
  type: ActionType;
  label: string;
  description: string;
  app: AppName;
  fields: FieldConfig[];
  summarize: (params: Record<string, unknown>) => string;
};

export const ACTIONS: ActionConfig[] = [
  {
    type: "send_email",
    label: "Send an email",
    description: "Compose and send a message through Gmail.",
    app: "Gmail",
    fields: [
      { name: "to", label: "To", kind: "email", placeholder: "someone@example.com" },
      { name: "subject", label: "Subject", kind: "text" },
      { name: "body", label: "Message", kind: "textarea" },
    ],
    summarize: (p) => `Send an email to ${p.to ?? "someone"}: "${p.subject ?? "(no subject)"}"`,
  },
  {
    type: "send_bulk_email",
    label: "Send a bulk email",
    description: "Personalize and send an email to everyone in an uploaded list.",
    app: "Gmail",
    fields: [
      { name: "recipients", label: "Recipient list", kind: "recipients" },
      { name: "subject", label: "Subject", kind: "text", placeholder: "Welcome {{name}}" },
      {
        name: "body",
        label: "Message",
        kind: "textarea",
        placeholder: "Hi {{name}}, welcome to Dispatch.",
      },
    ],
    summarize: (p) => {
      const count = Array.isArray(p.recipients) ? p.recipients.length : 0;
      return `Send a bulk email to ${count} recipient${count === 1 ? "" : "s"}: "${p.subject ?? "(no subject)"}"`;
    },
  },
  {
    type: "create_calendar_event",
    label: "Create a calendar event",
    description: "Add an event to Google Calendar.",
    app: "Google Calendar",
    fields: [
      { name: "title", label: "Title", kind: "text" },
      { name: "start_time", label: "Date & time", kind: "datetime-local" },
    ],
    summarize: (p) => `Create a calendar event: "${p.title ?? "Untitled event"}"`,
  },
  {
    type: "create_notion_page",
    label: "Create a Notion page",
    description: "Add a new page to Notion.",
    app: "Notion",
    fields: [
      { name: "title", label: "Title", kind: "text" },
      { name: "content", label: "Content", kind: "textarea" },
    ],
    summarize: (p) => `Create a Notion page: "${p.title ?? "Untitled page"}"`,
  },
];

export function actionFor(type: ActionType): ActionConfig {
  const action = ACTIONS.find((a) => a.type === type);
  if (!action) throw new Error(`Unknown action type: ${type}`);
  return action;
}

File: backend/app/services/action_registry.py
"""Action type definitions: the params schema and provider for every
``action_type`` a workflow step can have.

Lives apart from ``workflow_service.py`` so provider-specific integration
modules (``app.integrations.gmail``) can import these Pydantic models without
importing the workflow service itself — ``workflow_service`` depends on
``app.integrations`` (to look up an executor at run time), so the reverse
dependency would be circular.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

from app.models.enums import IntegrationProvider

# Direct-JSON bulk sends (not routed through the CSV/XLSX ingestion endpoint)
# are capped here — a business sending to more recipients should go through
# file upload, which is where real-world list sizes belong.
_MAX_BULK_RECIPIENTS = 500


class SendEmailParams(BaseModel):
    to: EmailStr
    subject: str = Field(min_length=1, max_length=300)
    body: str = Field(min_length=1, max_length=10000)


class CreateCalendarEventParams(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    start_time: datetime
    attendees: list[EmailStr] = Field(default_factory=list)


class CreateNotionPageParams(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    content: str = Field(min_length=1, max_length=10000)


class RecipientIn(BaseModel):
    """One bulk-send recipient. Any column beyond ``email`` (name, company,
    ...) is kept as-is and becomes a ``{{variable}}`` available when
    rendering that recipient's subject/body.
    """

    model_config = ConfigDict(extra="allow")

    email: EmailStr


class SendBulkEmailParams(BaseModel):
    recipients: list[RecipientIn] = Field(min_length=1, max_length=_MAX_BULK_RECIPIENTS)
    subject: str = Field(min_length=1, max_length=300)
    body: str = Field(min_length=1, max_length=10000)

    @field_validator("recipients")
    @classmethod
    def _reject_duplicate_emails(cls, recipients: list[RecipientIn]) -> list[RecipientIn]:
        seen: set[str] = set()
        duplicates: set[str] = set()
        for recipient in recipients:
            normalized = recipient.email.lower()
            if normalized in seen:
                duplicates.add(normalized)
            seen.add(normalized)
        if duplicates:
            raise ValueError(f"Duplicate recipient email(s): {', '.join(sorted(duplicates))}")
        return recipients


@dataclass(frozen=True)
class ActionSpec:
    label: str
    provider: IntegrationProvider
    params_model: type[BaseModel]


ACTION_REGISTRY: dict[str, ActionSpec] = {
    "send_email": ActionSpec(
        "Send an email",
        IntegrationProvider.GMAIL,
        SendEmailParams,
    ),
    "send_bulk_email": ActionSpec(
        "Send a bulk email",
        IntegrationProvider.GMAIL,
        SendBulkEmailParams,
    ),
    "create_calendar_event": ActionSpec(
        "Create a calendar event",
        IntegrationProvider.GOOGLE_CALENDAR,
        CreateCalendarEventParams,
    ),
    "create_notion_page": ActionSpec(
        "Create a Notion page",
        IntegrationProvider.NOTION,
        CreateNotionPageParams,
    ),
}
